flink-runtime/src/main/java/org/apache/flink/runtime/security/token/KerberosDelegationTokenManager.java

```diff
@@ -110,13 +126,84 @@ public void obtainDelegationTokens(Credentials credentials) {
      * task managers.
      */
     @Override
-    public void start() {
-        LOG.info("Starting renewal task");
+    public void start() throws Exception {
+        checkState(renewalExecutor == null, "Manager is already started");
+
+        if (!isRenewalPossible()) {
+            LOG.info("Renewal is NOT possible, skipping to start renewal task");
+            return;
+        }
+
+        ThreadFactory threadFactory =
+                new ThreadFactoryBuilder()
+                        .setDaemon(true)
+                        .setNameFormat("Credential Renewal Thread")
+                        .build();
+        renewalExecutor = new ScheduledThreadPoolExecutor(1, threadFactory);
```

Comment:
I'd love to avoid adding a new executor here (in general we try to avoid a thread pollution where possible). We should be able to reuse the existing executors that are already available to the `ResourceManager`:
- ComponentMainThreadExecutor (main thread for the ResourceManager endpoint; can be used for scheduling; can't have blocking runnables otherwise we'd disrupt the endpoint availability)
- ioExecutor (for blocking io heavy calls)

Basically the start method would become something along these lines:

```java
        final long tgtRenewalPeriod = configuration.get(KERBEROS_RELOGIN_PERIOD).toMillis();
        // The renewal future needs to be cancelled during close.
        final ScheduledFuture<?> renewalFuture =
                mainThreadExecutor.schedule(
                        () ->
                                ioExecutor.execute(
                                        () -> {
                                            // Todo...
                                        }),
                        tgtRenewalPeriod,
                        TimeUnit.MILLISECONDS);
                        ```

WDYT?

Reply:
I'm fine w/ using the already existing executors but I don't see any `ComponentMainThreadExecutor` in `ResourceManager`. Can you point where can I find it to pass to DTM?

Reply:
In the latest code `ComponentMainThreadExecutor` is null. When I know from where it should come from it will be filled.

Reply:
🤔 After little bit of digging, there is actually one more (scheduled) executor shared by the RPC services, that we could leverage. It should still be used in combination with the `ioExecutor`, because blocking there could potentially delay other critical things (eg. TM registration process).

Both `ioExecutor` and `rpcService.getScheduledExecutor()` are already available in the `ClusterEntrypoint` so it should be straightforward to pass them via constructor, without having to create additional factories.

Any thoughts?

```diff
diff --git a/flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypoint.java b/flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypoint.java
index 01c5f5f1c1c..8ee77e80d9c 100755
--- a/flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypoint.java
+++ b/flink-runtime/src/main/java/org/apache/flink/runtime/entrypoint/ClusterEntrypoint.java
@@ -395,7 +395,10 @@ public abstract class ClusterEntrypoint implements AutoCloseableAsync, FatalErro
                     configuration.getBoolean(SecurityOptions.KERBEROS_FETCH_DELEGATION_TOKEN)
                                     && HadoopDependency.isHadoopCommonOnClasspath(
                                             getClass().getClassLoader())
-                            ? new KerberosDelegationTokenManager(configuration)
+                            ? new KerberosDelegationTokenManager(
+                                    configuration,
+                                    commonRpcService.getScheduledExecutor(),
+                                    ioExecutor)
                             : new NoOpDelegationTokenManager();
             metricRegistry = createMetricRegistry(configuration, pluginManager, rpcSystem);
```

Reply:
That fully makes sense, `commonRpcService` also available in the `MiniCluster` area. Making the changes right now...

Reply:
I've tested it on cluster and works like charm so closing this discussion.